### PR TITLE
ss7ycIZj: Hash PIDs in the VSP

### PR DESCRIPTION
**Pull request by Rosa-Fox**
- We have left some comments on the `VerifyServiceProviderConfigurationTest.java` to make it easier for the reviewer to understand the `hashingEntityId` scenarios. We can remove these before merging.

- Adds the PID hashing functionality to the VSP as it will happen within here as opposed to the MSA for the non-matching journey.

- The VSP has support for multiple entity IDs (multi-tenanted). We therefore need to know which entity ID to use for hashing in the case of multi tenancy.

The longer version:

What it is?
Add hashing of the PIDs in the VSP as the functionality is currently in the MSA. The MSA currently uses its entity ID as part of the hashing algorithm. The VSP has support for multiple entity IDs (multi-tenanted) so we need to decide which one to hash with.

As most multi-tenanted organisations will want a user to have the same hashed PID (so that they have the same primary key in their user database), we will probably want to use a single entity ID from the VSP for hashing.

Decision: we will add a new parameter to the configuration file hashingEntityId that is used when doing the hashing. If there is only one serviceEntityId , hashingEntityId should default to that without the user have to specify it. If there are multiple serviceEntityIds, the user must specify the hashingEntityId

This will allow existing services that have a single MSA instance to specify their MSA entity ID as the hashing entity ID so historical users have their PIDs hashed in the same way.

Why do we need it?
As a Verify, we need to pass the identity to a connecting service and hash the PID so that privacy across the federation is preserved.

This PR was Co-authored-by: Olakunle Jegede <olakunle.jegede@digital.cabinet-office.gov.uk>

### CODE REVIEW

File: src/acceptance-test/java/uk/gov/ida/verifyserviceprovider/TranslateResponseMultiTenantedSetupAcceptanceTest.java [hunk 1]
```diff
@@ -75,7 +75,7 @@ public void shouldHandleASuccessMatchResponseForCorrectProvidedEntityId() {
 
     @Test
     public void shouldReturn400IfNoEntityIdProvided() {
-        complianceTool.initialiseWithEntityIdAndPid(configuredEntityIdOne, "some-expected-pid");
+        complianceTool.initialiseWithEntityIdAndPid(configuredEntityIdTwo, "some-expected-pid");
```
Comment (andy-paine) [inline src/acceptance-test/java/uk/gov/ida/verifyserviceprovider/TranslateResponseMultiTenantedSetupAcceptanceTest.java]: Do we need these changes?

Comment (Rosa-Fox) [reply src/acceptance-test/java/uk/gov/ida/verifyserviceprovider/TranslateResponseMultiTenantedSetupAcceptanceTest.java]: ah no we don't, have removed.

File: src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java [hunk 2]
```diff
@@ -16,48 +18,95 @@
...
-    @Valid
     private Duration clockSkew;
 
+    @JsonCreator
+    public VerifyServiceProviderConfiguration(
+        @JsonProperty("serviceEntityIds")
+        @NotNull
+        @Size(min = 1, message = NOT_EMPTY_MESSAGE)
+        @Valid
+        List<String> serviceEntityIds,
```
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java]: 👍 have fixed.

File: src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java [hunk 3]
```diff
@@ -16,48 +18,95 @@
...
     public List<String> getServiceEntityIds() {
         return serviceEntityIds;
     }
 
+    public String getHashingEntityId() {
+
+        if(hashingEntityId != null) {
+            return hashingEntityId;
+        }
+        if (getServiceEntityIds().isEmpty() == false && getServiceEntityIds().size() == 1) {
```
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java]: Thanks we have made this change, do you think we should include a test for if the service entity id is null?

File: src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java [hunk 4]
```diff
@@ -16,48 +18,95 @@
...
+            hashingEntityId = getServiceEntityIds().get(0);
+        }
+        if (hashingEntityId == null || hashingEntityId.isEmpty() && getServiceEntityIds().size() > 1) {
```
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java]: 👍 thanks, have changed.

File: src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java [hunk 5]
```diff
@@ -16,48 +18,95 @@
...
+            throw new NoHashingEntityIdIsProvidedError("No HashingEntityId is provided");
+        }
+        return hashingEntityId;
```
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java]: It will be reached by the result of line 59 which assigns `hashingEntityId`.

[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java]: The `return` is used to get the value of the assignment 
`hashingEntityId = getServiceEntityIds().get(0);`

File: src/test/java/unit/uk/gov/ida/verifyserviceprovider/VerifyServiceProviderConfigurationTest.java [hunk 6]
```diff
@@ -62,16 +73,80 @@ public void shouldNotComplainWhenConfiguredCorrectly() throws Exception {
...
+
+        VerifyServiceProviderConfiguration verifyServiceProviderConfiguration = asVerifyServiceProviderConfigurationCreator(
+                Arrays.asList("http://some-service-entity-id","http://some-service-entity-id2"),
+                null
+        );
+
+        expectedException.expect(NoHashingEntityIdIsProvidedError.class);
+        expectedException.expectMessage("No HashingEntityId is provided");
+
+        assertEquals(verifyServiceProviderConfiguration.getHashingEntityId(),"provided-hashing-entity-id");
```
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply src/test/java/unit/uk/gov/ida/verifyserviceprovider/VerifyServiceProviderConfigurationTest.java]: It should be, it catches the case where there are multiple `serviceEntityIds` and no `hashingEntityId` has been provided.

File: src/test/java/unit/uk/gov/ida/verifyserviceprovider/VerifyServiceProviderConfigurationTest.java [hunk 7]
```diff
@@ -62,16 +73,80 @@ public void shouldNotComplainWhenConfiguredCorrectly() throws Exception {
...
+    //when hashing entity id is defined, return it
+    @Test
+    public void shouldReturnHashingEntityIdWhenItIsDefined() {
+
+        VerifyServiceProviderConfiguration verifyServiceProviderConfiguration = asVerifyServiceProviderConfigurationCreator(
+                Arrays.asList("http://some-service-entity-id","http://some-service-entity-id2"),
+                "provided-hashing-entity-id"
+        );
+
+        assertEquals(verifyServiceProviderConfiguration.getHashingEntityId(),"provided-hashing-entity-id");
```
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply src/test/java/unit/uk/gov/ida/verifyserviceprovider/VerifyServiceProviderConfigurationTest.java]: The closest we got was `assertThat(verifyServiceProviderConfiguration.getHashingEntityId()).isEqualTo("provided-hashing-entity-id");`. Does read better :)

File: src/acceptance-test/java/uk/gov/ida/verifyserviceprovider/TranslateResponseMultiTenantedSetupAcceptanceTest.java (hunk 1 above)
Comment (Rosa-Fox) [reply src/acceptance-test/java/uk/gov/ida/verifyserviceprovider/TranslateResponseMultiTenantedSetupAcceptanceTest.java]: Have removed them.

File: src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java]: Have amended.

File: src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java]: 👍

File: src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply src/main/java/uk/gov/ida/verifyserviceprovider/configuration/VerifyServiceProviderConfiguration.java]: 👍

File: src/test/java/unit/uk/gov/ida/verifyserviceprovider/VerifyServiceProviderConfigurationTest.java [hunk 8]
```diff
@@ -35,9 +45,10 @@
     );
     private EnvironmentHelper environmentHelper = new EnvironmentHelper();
 
+
     @After
```
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply src/test/java/unit/uk/gov/ida/verifyserviceprovider/VerifyServiceProviderConfigurationTest.java]: 👍

File: src/test/java/unit/uk/gov/ida/verifyserviceprovider/VerifyServiceProviderConfigurationTest.java [hunk 9]
```diff
@@ -62,16 +73,80 @@ public void shouldNotComplainWhenConfiguredCorrectly() throws Exception {
...
+                null
+        );
+
+        expectedException.expect(NoHashingEntityIdIsProvidedError.class);
+        expectedException.expectMessage("No HashingEntityId is provided");
+
+        assertEquals(verifyServiceProviderConfiguration.getHashingEntityId(),"provided-hashing-entity-id");
+    }
+
+    private VerifyServiceProviderConfiguration asVerifyServiceProviderConfigurationCreator(List<String> serviceEntityIds, String hashingEntityId) {
```
[the remark this answers is not in the document]
Comment (Rosa-Fox) [reply src/test/java/unit/uk/gov/ida/verifyserviceprovider/VerifyServiceProviderConfigurationTest.java]: 👍
